docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,366 @@
+BP003
+#####
+
+:Number: BP003
+:Title: N(ext)Runner Task Scheduler
+:Author: Cleber Rosa <crosa@redhat.com>
+:Discussions-To: avocado-devel@redhat.com
+:Reviewers:
+:Created: 20-July-2020
+:Type: Architecture Blueprint
+:Status: Draft
+
+.. contents:: Table of Contents
+
+TL;DR
+*****
+
+The N(ext) Runner has been used as Avocado's runner for selftests for
+over a year, but that has been implemented as ``avocado nrun``, that
+is, outside of the Avocado's traditional ``avocado run`` entrypoint.
+Under the hood, it means that the N(ext) Runner is not well integrated
+enough to an Avocado Job.
+
+A partial implementation of the N(ext) Runner integration to an
+Avocado Job is available at ``avocado/plugins/runner_nrunner.py`` but
+it has many limitations.  This blueprint describes the architecture of
+a scheduler that will power this Job compatible N(ext) Runner
+implementation.
+
+Motivation
+**********
+
+Describe the architecture of an task scheduler that will power the Job
+compatible N(ext) Runner implementation, allowing the N(ext) Runner
+first supported version to achieve feature completeness, and replace
+the current runner soon after the next LTS version is released.
+
+Specification
+*************
+
+Task execution coordination goals
+---------------------------------
+
+For an Avocado job to be useful with the N(ext) Runner architecture,
+tasks must be executed. Differently then the current runner, the
+N(ext) Runner architecture allows those to be executed in a much more
+decoupled way. This characteristic will be maintained, but it needs to
+be adapted into the current Job execution.
+
+From a high-level view, the nrunner plugin needs to:
+
+1. Break apart from the "one at a time" Task execution model that it currently
+   employs;
+
+2. Check if a Tasks can be executed, that is, if its requirements can be
+   fulfilled (the most basic requirement for a task is a matching runner);
+
+3. Prepare for the execution of a task, such as the fulfillment of extra task
+   requirements. The requirements resolver is one, if not the only way,
+   component that should be given a chance to act here;
+
+4. Executes a task in a prepared environment;
+
+5. Monitor the execution of a task (from an external PoV);
+
+6. Collect the status messages that tasks will send;
+
+  a. Forward the status messages to the appropriate job components, such as the
+  result plugins.
```

Comment:
_"I don't see the scheduler as a bridge between the task and the job."_

Again, this is probably the same bad name choice I made.  I'm confident that something doing that necessary though.

_" I think the task should communicate directly with the job."_

Even today, a task may be running on a container with the Podman spawner.  So how do you propose this "direct communication"?

My take: a task may be running elsewhere, and may not have direct access to the Job (meaning an instance of `avocado.core.job.Job`).  So *something* (with a good name) needs to act as an intermediary.

_"I see just "checkpoint" communication happening between the task and the scheduler, like "I'm still alive and running," or "I'm done."_

So, you consider a message such as "I'm done" a "checkpoint communication", while I called it "status messages", which seems to me different names for the same thing.  But, you propose that this communication "happening between the task and the scheduler" never leaves the scheduler. So, do you think the task should send the same (or very similar) information to the scheduler and *also* to the Job (given that you "don't see the scheduler as a bridge between the task and the job") ?

_"In the reverse flow, from the scheduler to the task, "You reached your allowed time to run."_

I haven't come across the need to have information flowing this way, which is good because it would add a LOT of complexity in the components running the tasks (the current `avocado-runner-*` scripts).  When a situation such as a timeout happens, the "scheduler" (or whatever name it's given) can attempt to destroy the environment (process, container, etc) and collect the (partial?) data.

Reply:
> _"I don't see the scheduler as a bridge between the task and the job."_
> 
> Again, this is probably the same bad name choice I made. I'm confident that something doing that _is_ necessary though.

If the implementation needs a single point of management and communication, we may be talking about a scheduler. I really think we don't need the task talking to a centralized point of management. It may talk to a node in the state machine, and that is acceptable.

> 
> _" I think the task should communicate directly with the job."_
> 
> Even today, a task may be running on a container with the Podman spawner. So how do you propose this "direct communication"?
> 
> My take: a task may be running elsewhere, and may not have direct access to the Job (meaning an instance of `avocado.core.job.Job`). So _something_ (with a good name) needs to act as an intermediary.
> 

I think I got what you mean and maybe we are thinking about the same thing. I would say the status server is a great candidate for that communication. And what I meant was the Job receiving communication thru the status server.

Anyway, I think this is an implementation detail that is not covered here right not.

> _"I see just "checkpoint" communication happening between the task and the scheduler, like "I'm still alive and running," or "I'm done."_
> 
> So, you consider a message such as "I'm done" a "checkpoint communication", while I called it "status messages", which seems to me different names for the same thing. But, you propose that this communication "happening between the task and the scheduler" never leaves the scheduler. So, do you think the task should send the same (or very similar) information to the scheduler and _also_ to the Job (given that you "don't see the scheduler as a bridge between the task and the job") ?
> 

No, the scheduler do not send anything to the Job. If we talk strictly about a scheduler here, it is the scheduler's responsibility to manage the flow of tasks. The Job can query for the status if it is desired, but it is not the scheduler's responsibility to actively send it to the Job. If we talk about something else that is not a scheduler, there can be communication between the task and the node of a state machine, for example, and also between the Job and the node of the state machine.

> _"In the reverse flow, from the scheduler to the task, "You reached your allowed time to run."_
> 
> I haven't come across the need to have information flowing this way, which is good because it would add a LOT of complexity in the components running the tasks (the current `avocado-runner-*` scripts). When a situation such as a timeout happens, the "scheduler" (or whatever name it's given) can attempt to destroy the environment (process, container, etc) and collect the (partial?) data.

Okay, I was thinking about a "soft" task cancel. If that is not needed, reverse communication is not needed.